src/hazardous/secret_stream/xchacha20poly1305.rs

```diff
@@ -0,0 +1,454 @@
+// all copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// Based on the implementation used by libsodium
+// https://download.libsodium.org/doc/secret-key_cryptography/secretstream
+// https://github.com/jedisct1/libsodium/blob/stable/src/libsodium/crypto_secretstream/xchacha20poly1305/secretstream_xchacha20poly1305.c
+
+use crate::errors::UnknownCryptoError;
+use crate::hazardous::mac::poly1305::{init, OneTimeKey, POLY1305_KEYSIZE, POLY1305_OUTSIZE};
+use crate::hazardous::stream::chacha20::{
+	encrypt as chacha20_enc, encrypt_in_place as chacha20_enc_in_place, hchacha20,
+	Nonce as chacha20Nonce, SecretKey as chacha20Key, CHACHA_KEYSIZE, HCHACHA_NONCESIZE,
+	IETF_CHACHA_NONCESIZE,
+};
+pub use crate::hazardous::stream::xchacha20::Nonce;
+
+use crate::util::secure_cmp;
+use bitflags::bitflags;
+
+//TODO Add more tests
+
+//TODO: Im not sure if this is the right place for the bit flags and if they are needed at all.
+// It would also be possible to only define the REKEY value and pass the tag around as u8 and define
+// the user friendly bitflags/enum in the safe API. Im not sure how to work around using the
+// bitflags create. rust enums are rather limited in functionality as flags and im not a fan of
+// passing around raw u8 bitflags.
+bitflags! {
+	/// Tag that indicates the type of message
+	pub struct Tag: u8 {
+		/// A simple message with no special meaning
+		const MESSAGE = 0b0000_0000;
+		/// Marks that the message is the end of a set of messages. Allows the decrypting site to
+		/// start working with this data
+		const PUSH = 0b0000_0001;
+		/// derives a new secret key and forgets the one used for earlier encryption/decryption
+		const REKEY = 0b0000_0010;
+		/// Indicates the end of a stream. Also does an rekey.
+		const FINISH = Self::PUSH.bits | Self::REKEY.bits;
+
+	}
+}
+
+/// Size of the header/nonce
+pub const SECRETSTREAM_XCHACHA20POLY1305_HEADERBYTES: usize = 24;
+const SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES: usize = 4;
+const SECRETSTREAM_XCHACHA20POLY1305_INONCEBYTES: usize = 8;
+/// Size of additional data appended to each message
+pub const SECRETSTREAM_XCHACHA20POLY1305_ABYTES: usize =
+	POLY1305_OUTSIZE + core::mem::size_of::<Tag>();
+//TODO assert that counterbytes + inoncebytes = IETF_CHACHA_NONCESIZE
+
+fn xor_buf8(out: &mut [u8], input: &[u8]) {
+	debug_assert_eq!(out.len(), 8);
+	debug_assert_eq!(input.len(), 8);
+	for i in out.iter_mut().zip(input.iter()) {
+		*i.0 ^= *i.1;
```

Comment:
NIT: Instead of combining these two I think the loop could be made clearer by splitting them up:
```
for (o_elem, i_elem) in out.iter_mut().zip(input.iter()) {
		*o_elem ^= i_elem;
```